### ISSUE TITLE
Blazor - support variable height items in Virtualize component

### ISSUE THREAD

**Issue by radderz** [dotnet/aspnetcore#25058]
When create a listview of **similarly** sized items, it's very hard to virtualise this in most web frameworks currently. It'd be great for Blazor to provide an out of the box solution that is able to render a virtual list of items, and account for it's position in the list as items are rendered. Most of my implementations of lists are of items that don't vary in size all that greatly, i.e. an extra line or two of text per item due to some wrapping.

This would need to be able to handle data updates over time, so it shouldn't remember the size of the item when it goes back out of the view port and if the item is in the viewport it should account for the change in size after an update while assuming an approximate size for the not currently visible items.

My concept is to try to have the scroll bar managing the item number through the list rather than the absolute x position, then to show the item at the top that matches closest to that by %, if snapping the items makes this easier that may be an approach.

Basically you'd know the items sizes currently in the viewport, and as the user scrolls, the overall size of the view port would adjust based on the currently visible items actual size + the amount of items above * approx size + the amount of items below * approx size.

This is done well in Xamarin, but that is a non web framework so I guess it's a little easier to control the outcome.

**Comment by mrpmorris** [comment]
How could you know the sizes of items not currently in the DOM?

**Comment by mrlife** [comment]
For variable height items (e.g. due to text wrapping), if we could keep visible items from "jumping" when new items are rendered, that would be really helpful. As of now, if I'm reading a visible item and new items render, the item I'm reading moves, causing a disorienting effect.

**Comment by legistek** [comment]
> How could you know the sizes of items not currently in the DOM?
Render them in a hidden DIV off screen and measure them.

Also see the part of the best Microsoft framework ever made for inspiration:
https://github.com/dotnet/wpf/blob/master/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/VirtualizingStackPanel.cs

**Comment by radderz** [comment]
@legistek I haven't seen any good component that implements this from the component vendors like Telerik, Syncfusion etc on their web frameworks, this would make Blazor really compelling to me as it would remove the biggest pain point I have with HTML5, if the base virtualize container was capable of this, it'd make a lot of UIs and other components much easier to build while maintaining that benefit of virtualizing the large lists.

It is base level functionality of a virtual scrolling system on mobile but I do understand that they are running native UI which enables more control over the resulting scroll behaviour.

**Comment by legistek** [comment]
I recently discovered the Microsoft Fluent UI React controls and they support virtualization with arbitrary and variable row heights:

https://developer.microsoft.com/en-us/fluentui#/controls/web/detailslist/variablerowheights

It's quite good. Not *as good* as the virtualizing panels in the native Windows XAML platforms but still very good. I haven't delved into the code to see how they work but by definition if it can be done in React it could be done in Blazor.

**Comment by masoodafar-web** [comment]
Hi, do you have a plan for this in the future?
Because our friend claims that they have solved this problem for react.

**Comment by LuohuaRain** [comment]
I have the same question

**Comment by GioviQ** [comment]
This implementation https://www.meziantou.net/infinite-scrolling-in-blazor.htm seems better.

**Comment by realivanjx** [comment]
@GioviQ infinite scrolling is not the same as virtualization

**Comment by GioviQ** [comment]
@realivanjx can you detail the differences, because to me both load items on demand.

**Comment by neckaros** [comment]
> @realivanjx can you detail the differences, because to me both load items on demand.

Virtualized only render in the DOM a subset of data available. 
infinite scroll load data from server as you scroll but add all loaded data in the DOM

**Comment by rf-0** [comment]
The height limitation (ItemSize), unfortunately, is quite a big limitation as mentioned above and feels more like a workaround solution.

This is a much needed feature for one of the most important components for the current and future web. Looking forward for this to be implemented in the future. Nice work so far to all the Blazor team.

**Comment by wrkntwrkn** [comment]
Hi, this is a very important feature for reasons mentioned above among others. Hopefully this will be prioritized.

**Comment by mrpmorris** [comment]
I am guessing the solution to this would be something like...

[Page 1]
<[Page 2]>
[Page 3]

When the first item on [Page3] becomes visible you do the following
1: Move P2 into P1
2: Move P3 into P2
3: Fetch P3 data

Very idle thoughts, most likely should be ignored :)

**Comment by ForeverFast** [comment]
Is there any progress or work on this functionality?

**Comment by rf-0** [comment]
Any update on this? Its a much needed feature.

**Comment by gldraphael** [comment]
According to the labels, this issue is not being considered for .NET 7.
Our best bet is to upvote this issue, and share specific use-cases as a comment.

**Comment by rf-0** [comment]
In terms of use cases, some were mentioned above. But in today's internet age, most content is scrollable content and having this limitation is a problem.

The reason why a workaround wont be enough is because, even after following the performance guidelines for blazor webassembly (standalone) from microsoft, it will still be a problem in terms of cpu/mem usage when having more than a few hundred items loaded, even when their UI is not too complex. (Pagination or other methods not an option).

**Comment by koureasstavros** [comment]
I have a huge list of tables while each table has it's own fields. I also have an option to expand or collapse the fields. When the fields are collapsed and there are only rows with tables, the height of each row is the same, and everything is working as expected. On the other side, when the fields are expanded the height of each row is different, and causes strange scrolling issues. For example when more items are loaded the page scrolls to previous rows I have already passed.

In my case I have already all the tables and the fields in memory (in variable), so based on my css I could know the height of each row, for example RowHeight="@(table.fields.count() * 25)px;"

**Comment by thirstyape** [comment]
I think that @SteveSandersonMS has provided an excellent answer in [this comment](https://github.com/dotnet/aspnetcore/issues/42518#issuecomment-1173609044) detailing much of the reasoning behind why this and similar issues may be difficult to tackle.

That said, I agree with @rf-0 in that this limitation is a problem. Think of a simple data card containing a title, an image, and some text. While they will all have a relatively similar size there may be slight variance due to the size of the image or amount of text. 

Speaking to Blazor WebAssembly specifically, one of the really great things about it is you can scale an app with minimal server requirements by having a backend API while the user device takes care of the rendering and UI work. This allows running an app on almost any device. But this limitation of not being able to virtualize large data sets can easily consume all of the memory on older or less powerful devices (think of a certain online retailer selling tablets using the technology of yesteryear).

I think one possible solution, albeit not very elegant, would be to allow the `ItemSize` parameter to be used as an average item size. This would still provide the ability to have a near fixed size scrollbar. With some slight adjustments to the scrollbar size depending on the actual size of the currently loaded items. Thus the scroll position would still determine the start index and the amount of items loaded would be based on the average item size. Combined with the `OverscanCount` parameter I think this would be not too difficult to implement and add a great deal to the usability of the component.

I will say this, please, please, please remove the `sealed` keyword from [Virtualize](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web/src/Virtualization/Virtualize.cs) so that I, and I'm sure many others, can try these ideas out. Why bother having an open-source framework and make it intentionally difficult to extend?

**Comment by wrkntwrkn** [comment]
@mrpmorris If you do have a solution please do share. We use infinite scrolling (have our own custom component) in quite a few areas of the web app we are currently building (also using your **[Fluxor](https://github.com/mrpmorris/Fluxor)** library. brilliant library btw! thanks!) and this would be a match in heaven! At the moment loading/rendering so many items into comes with quite a few performance costs when using webassembly, especially when navigating away to some other page or rerendering the items already loaded. Obviously, this isn't the case when using Virtualize which would be the ideal component for what we want to do. @thirstyape solution could work as well. Being able to precompute the size of each item and let the Virtualize component know in advance before rendering could potentially work? We could precompute this easily based on the content. So in this case the Item size would be defined on the child component, per item, instead at the parent level (Virtualize).

**Comment by szalapski** [comment]
I came across the need for variable-height items today as well. Would love to see it "just work", though I understand the challenges noted above.

**Comment by adalipi** [comment]
I tried to implement something like unlimited scroll with variable length of items per row and no matter the style of items, it is more like a hack rather than nicely done but it works for me it might just help you guys what virtualization is missing. Check it on stack overflow: [here](https://stackoverflow.com/questions/72581334/in-blazor-when-using-the-virtualize-component-with-multiple-items-per-row-is-th/77987719#77987719)

**Comment by koureasstavros** [comment]
Hi, any update on this topic with the variable height of items while using virtualize?

**Comment by tripolskypetr** [comment]
There is a high possibility of creator of this issue died from covid, but the bot continues to reschedule resolvement of that issue

![image](https://github.com/dotnet/aspnetcore/assets/19227776/1166d71d-038c-45ae-97eb-fb96d6c899c0)

![image](https://github.com/dotnet/aspnetcore/assets/19227776/4eef7623-9821-4f02-8c3c-31a711bf5d78)

![image](https://github.com/dotnet/aspnetcore/assets/19227776/a2ef0ed9-b7e4-4770-a39e-f98af6fbd58e)

**Comment by radderz** [comment]
Nope, still alive. This issue is a feature request and it hasn't been prioritised by Microsoft. I went a different direction with paging since there was no commitment to enable such a feature.

**Comment by tripolskypetr** [comment]
The paging is not good enough cause when using firebase you are forced to use cursor pagination.

**Comment by emcnickle** [comment]
This would be very useful even with the known complications talked about above. We use a lot of telerik controls and they are reluctant to take it on either for the same reasons all of which would be acceptable compromises for the performance gain in given scenarios.

**Comment by KieranDevvs** [comment]
Glad to see this might get prioritised for .NET 11, its a real big pain in Blazor at the moment for me 👍🏻

**Comment by radderz** [comment]
If it does get prioritised, hopefully it will be done in a way where the item sizes can change over time (think updating of items) without that causing havoc, maybe the scroll bar would need to be virtualized so that it is item count centric rather than absolute position based, i.e. scrolling to the same point would always show the same topmost item in view no matter how large the items above it are.

**Comment by ilonatommy** [comment]
This issue contains 3 feature asks that could be grouped as follows:
- **Core variable-height virtualization** — Implement estimation-based sizing for off-screen items and measure actual heights as items enter the viewport. Row-increment scrolling is an acceptable simplification for the initial implementation.

- **Item-count-centric scrollbar** — Replace pixel-position-based scrolling with item-index-based positioning, so scrolling to the same position always shows the same topmost item regardless of varying heights above it.

- **Dynamic content handling** — Support re-measuring and re-layout when item content changes while visible, without breaking scroll position or causing visual jank.

**Comment by ilonatommy** [comment]
## Initial core variable-height virtualization plan

The `Virtualize` component will automatically handle items of varying heights without any code changes.

How it would work:
Initial render: Items are rendered using the `ItemSize` parameter (default 50px) as an estimated height.
As items become visible: The component automatically measures their actual heights and caches them.
Subsequent scrolls: Measured heights are used for accurate positioning; unmeasured items still use the estimate.
Over time: As more items are measured, the scrollbar and positioning become increasingly accurate.

| Scenario | Target Experience |
|----------|------------|
| Items all same height | Identical to current behavior |
| Items vary slightly (±1-2 lines) | Smooth scrolling with minor adjustments |
| Items vary significantly | Smooth scrolling, but may see occasional small shifts as heights are corrected |

| Scenario | Target Approach | Trade-off |
|----------|--------------|-----------|
| User scrolls gradually | Measurements accumulate, positioning is accurate |  Works well |
| User jumps via scrollbar | Use estimates (`ItemSize` or average), render best guess | May show wrong items initially, then correct |
| User jumps, then scrolls slightly | Measurements kick in, position corrects | Brief visual adjustment |

### What We **Cannot** Solve:
- **Perfect positioning on jump** — impossible without measuring all items
- **Stable scrollbar size before measuring** — the thumb size will adjust as items are measured

**Comment by nathanaelfrey** [comment]
Great to see some movement here.

I assume that a scenario such as **User presses Ctrl + Home or Ctrl + End** would be handled the same as **User jumps via scrollbar**? From my observations using `<Virtualize ItemsProvider="MyItemsProvider" />` both of those scenarios tend to have several rapid and seemingly random calls to the provider function followed by a call for the desired items.

**Comment by ilonatommy** [comment]
> I assume that a scenario such as **User presses Ctrl + Home or Ctrl + End** would be handled the same as **User jumps via scrollbar**? From my observations using `<Virtualize ItemsProvider="MyItemsProvider" />` both of those scenarios tend to have several rapid and seemingly random calls to the provider function followed by a call for the desired items.

Jumping will be supported. Thank you for reporting the issue, it's indeed present with a custom item provider. It looks likely it has the same root cause as https://github.com/dotnet/aspnetcore/issues/64029, I will link it there.

**Comment by radderz** [comment]
One important thing to consider:

When scrolling through a list of items that are variably sized, and using the scroll thumb, the scrolling goes crazy due to the item lengths changing and it can get very easily in a loop, so it needs to be stable during continuous scrolling even if the items are not perfectly scrolled to. 

Generally virtualization is used when you likely have many items in the list so the thumb is quite small, and dragging it around is all kinds of crazy with the current implementation when having variable sized items.

So a requirement would be that when scrolling with the thumb, even if the thumb size changes as items are added/removed from the dom, the scroll itself remains stable during the single continuous scroll. I think it's less important that the exact position in the list is guaranteed based on a specific scroll point.
